File: js/packages/teams-ai/src/prompts/DataSourceSection.ts
/**
 * @module teams-ai
 */
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */

import { Message } from './Message';
import { PromptFunctions } from './PromptFunctions';
import { RenderedPromptSection } from './PromptSection';
import { PromptSectionBase } from './PromptSectionBase';
import { TurnContext } from 'botbuilder';
import { Tokenizer } from '../tokenizers';
import { DataSource } from '../dataSources';
import { Memory } from '../MemoryFork';

/**
 * A section that renders a data source to a prompt.
 */
export class DataSourceSection extends PromptSectionBase {
    private readonly _dataSource: DataSource;

    /**
     * Creates a new `DataSourceSection` instance.
     * @param {DataSource} dataSource - The data source to render.
     * @param {number} tokens - Desired number of tokens to render.
     */
    public constructor(dataSource: DataSource, tokens: number) {
        super(tokens, true, '\n\n');
        this._dataSource = dataSource;
    }

    /**
     * @private
     * @param {TurnContext} context - Context for the current turn of conversation.
     * @param {Memory} memory - Memory to use for rendering.
     * @param {PromptFunctions} functions - Prompt functions to use for rendering.
     * @param {Tokenizer} tokenizer - Tokenizer to use for encoding text.
     * @param {number} maxTokens - Maximum number of tokens allowed.
     * @returns {Promise<RenderedPromptSection<Message<string>[]>>} Rendered prompt section as a string.
     */
    public async renderAsMessages(
        context: TurnContext,
        memory: Memory,
        functions: PromptFunctions,
        tokenizer: Tokenizer,
        maxTokens: number
    ): Promise<RenderedPromptSection<Message<string>[]>> {
        // Render data source
        const budget = this.getTokenBudget(maxTokens);
        const rendered = await this._dataSource.renderData(context, memory, tokenizer, budget);

        // Return as a 'system' message
        // - The role will typically end up being ignored because as this section is usually added
        //   to a `GroupSection` which will override the role.
        return {
            output: [{ role: 'system', content: rendered.output }],
            length: rendered.length,
            tooLong: rendered.tooLong
        };
    }
}

File: python/packages/ai/teams/ai/prompts/sections/data_source_section.py
"""
Copyright (c) Microsoft Corporation. All rights reserved.
Licensed under the MIT License.
"""

from __future__ import annotations

from typing import List

from botbuilder.core import TurnContext

from ....state import MemoryBase
from ...data_sources import DataSource
from ...tokenizers import Tokenizer
from ..message import Message
from ..prompt_functions import PromptFunctions
from ..rendered_prompt_section import RenderedPromptSection
from .prompt_section_base import PromptSectionBase


class DataSourceSection(PromptSectionBase):
    """
    A section that renders a data source to a prompt.
    """

    _data_source: DataSource

    def __init__(self, data_source: DataSource, tokens: float):
        """
        Creates a new `DataSourceSection` instance.

        Args:
            dataSource (DataSource): The data source to render.
            tokens (int): Sizing strategy for this section.
        """
        super().__init__(tokens, True, "\n\n")
        self._data_source = data_source

    async def render_as_messages(
        self,
        context: TurnContext,
        memory: MemoryBase,
        functions: PromptFunctions,
        tokenizer: Tokenizer,
        max_tokens: int,
    ) -> RenderedPromptSection[List[Message[str]]]:
        """
        Renders the section as a list of messages.

        Args:
            context (TurnContext): Context for the current turn of conversation with the user.
            memory (MemoryBase): An interface for accessing state values.
            functions (PromptFunctions): Registry of functions that can be used by the section.
            tokenizer (Tokenizer): Tokenizer to use when rendering the section.
            max_tokens (int): Maximum number of tokens allowed to be rendered.

        Returns:
            RenderedPromptSection[List[Message]]: The rendered prompt section as a list of messages.
        """
        # Render data source
        budget = self._get_token_budget(max_tokens)
        rendered = await self._data_source.render_data(context, memory, tokenizer, budget)

        # Return as a 'system' message
        # - The role will typically end up being ignored because as this section is usually added
        #   to a `GroupSection` which will override the role.
        return RenderedPromptSection(
            output=[Message[str]("system", rendered.output)],
            length=rendered.length,
            too_long=rendered.too_long,
        )
